File: packages/headless/DocJsonParser.rb
require 'json'

DOC_GEN_PATH = './dist/doc.json'
CONFIG_PATH = './docs.config.json'
OUTPUT_PATH = './dist/parsed_doc.json'

class DocJsonParser
  def initialize(doc_gen_path, config_path)
    @doc_gen = JSON.parse(File.read(doc_gen_path))
    @config = JSON.parse(File.read(config_path))
  end

  def parse
    {
      'engine' => parse_engine,
      'controllers' => parse_controllers,
      'actions' => parse_actions
    }
  end

  private

  # Methods to retrieve stuff from the docgen

  def get_modules(module_path)
    @doc_gen['children'].select do |child|
      child['originalName'].include?(module_path)
    end
  end

  def get_entity_from_modules_by_name(modules, name)
    modules.each do |mod|
      entity = get_entity_from_module_by_name(mod, name)
      return entity unless entity.nil?
    end
    return nil
  end

  def get_entity_from_module_by_name(mod, name)
    mod['children'].find do |object|
      object['name'] == name
    end
  end

  def get_entities_from_modules_by_kind_string(modules, kind_string)
    entities = []
    modules.each do |mod|
      entities.concat(get_entities_from_module_by_kind_string(mod, kind_string))
    end
    entities
  end

  def get_entities_from_module_by_kind_string(mod, kind_string)
    mod['children'].select do |object|
      object['kindString'] == kind_string
    end
  end

  def get_desc(comment)
    return '' unless comment.respond_to?('key?')

    unless comment.key?('shortText') && comment.key?('text')
      return (comment.key?('shortText') ? comment['shortText'] : comment['text']).to_s
    end

    "#{comment['shortText']}\n\n#{comment['text']}"
  end

  def get_returns(comment)
    return '' unless comment.respond_to?('key?')

    comment.key?('returns') ? comment['returns'] : ''
  end

  # ----------------------Type Resolver-----------------------

  def get_type(type_object, section)
    return {} if type_object.nil?
    return process_string_literal(type_object) if is_type?(type_object, 'stringLiteral')
    return process_intrinsic_type(type_object) if is_type?(type_object, 'intrinsic') || is_type?(type_object, 'unknown')
    return process_array_type(type_object) if is_type?(type_object, 'array')
    return process_union_type(type_object, section) if is_type?(type_object, 'union')
    return process_intersection_type(type_object) if is_type?(type_object, 'intersection')
    return process_reference_type(type_object, section) if is_type?(type_object, 'reference')
    return process_type_parameter_type(type_object) if is_type?(type_object, 'typeParameter')
    return process_reflection_type(type_object) if is_type?(type_object, 'reflection')

    {}
  end

  def is_type?(type_object, target_type)
    type_object['type'] == target_type
  end

  def process_string_literal(type_object)
    {
      'type' => type_object['type'],
      'value' => type_object['value']
    }
  end

  def process_intrinsic_type(type_object)
    {
      'type' => type_object['type'],
      'name' => type_object['name']
    }
  end

  def process_array_type(type_object)
    {
      'type' => 'array',
      'arrayOf' => get_type(type_object['elementType'], 'array')
    }
  end

  def process_union_type(type_object, section)
    processed_type = {
      'type' => 'union',
      'unionOf' => type_object['types'].map do |type|
        get_type(type, 'union')
      end
    }
    if section == 'interface' && processed_type['unionOf'].include?({'type' => 'intrinsic', 'name' => 'undefined'})
      processed_type['optional'] = 'true'
      processed_type['unionOf'].reject! { |type| type == {'type' => 'intrinsic', 'name' => 'undefined'}}
    end
    processed_type
  end

  def process_intersection_type(type_object)
    processed_type = {
      'type' => 'intersection',
      'intersectionOf' => type_object['types'].map do |member|
        get_type(member, 'intersection')
      end
    }
  end

  def process_reference_type(type_object, section)
    processed_type = {
      'type' => 'reference',
      'name' => type_object['name']
    }
    if type_object.key?('typeArguments')
      processed_type['type_parameters'] = type_object['typeArguments'].map do |type_arg|
        get_type(type_arg, section)
      end
    end
    processed_type
  end

  def process_type_parameter_type(type_object)
    processed_type = {
      'type' => 'type_parameter',
      'name' => type_object['name']
    }
    if type_object.key?('constraint')
      processed_type['extends'] = get_type(type_object['constraint'], 'generic')
    end
    if type_object.key?('default')
      processed_type['default'] = get_type(type_object['default'], 'generic')
    end
    processed_type
  end

  def process_reflection_type(type_object)
    if type_object['declaration'].key?('children')
      {
        'type' => 'object',
        'properties' => type_object['declaration']['children'].map do |property|
          {
            'name' => property['name'],
            'type' => get_type(property['type'], 'object')
          }
        end
      }
    elsif type_object['declaration'].key?('signatures')
      processed_type = {
        'type' => 'function',
        'parameters' => [],
        'returns' => get_type(type_object['declaration']['signatures'].first['type'], 'function')
      }
      if type_object['declaration']['signatures'].first.key?('parameters')
        type_object['declaration']['signatures'].first['parameters'].each do |param|
          processed_type['parameters'].push({
            'name' => param['name'],
            'type' => get_type(param['type'], 'parameter')
          })
        end
      end
      processed_type
    else
      {}
    end
  end

  # Expand the various entities

  def alphabetize(object_array)
    object_array.sort! do |a, b|
      a['name'] <=> b['name']
    end
  end

  def expand_functions(functions)
    expand_any_functions(functions, 'Call signature', 'signatures')
  end

  def expand_accessors(functions)
    expand_any_functions(functions, 'Get signature', 'getSignature')
  end

  def expand_constructor(function)
    expand_any_functions([function], 'Constructor signature', 'signatures').first
  end

  def expand_types(types)
    types_expanded = types.map do |type|
      type_object = {
        'name' => type['name'],
        'type' => get_type(type['type'], 'type'),
        'text' => get_desc(type['comment'])
      }
      type_object['type_parameters'] = expand_type_parameters(type['typeParameter']) if type.key?('typeParameter')
      type_object
    end
    alphabetize(types_expanded)
    types_expanded
  end

  def expand_interfaces(interfaces)
    interfaces_expanded = []
    interfaces.each do |interface|
      interface_object = {
        'name' => interface['name'],
        'text' => get_desc(interface['comment']),
        'properties' => []
      }
      interface_object['type_parameters'] = expand_type_parameters(interface['typeParameter']) if interface.key?('typeParameter')
      interface['children'].each do |property|
        property_object = {
          'name' => property['name'],
          'type' => get_type(property['type'], 'interface'),
          'text' => get_desc(property['comment'])
        }
        property_object['optional'] = 'true' if property['flags'].key?('isOptional')
        interface_object['properties'].push(property_object)
      end
      alphabetize(interface_object['properties'])
      interfaces_expanded.push(interface_object)
    end
    alphabetize(interfaces_expanded)
    interfaces_expanded
  end

  def expand_enums(enums)
    enums_expanded = []
    enums.each do |enum|
      enum_object = {
        'name' => enum['name'],
        'text' => get_desc(enum['comment']),
        'members' => []
      }
      enum['children'].each do |member|
        member_object = {
          'name' => member['name'],
          'text' => get_desc(member['comment'])
        }
        enum_object['members'].push(member_object)
      end
      alphabetize(enum_object['members'])
      enums_expanded.push(enum_object)
    end
    alphabetize(enums_expanded)
    enums_expanded
  end

  def expand_any_functions(functions, kind_string, signature_property)
    functions_expanded = []
    functions.each do |function|
      function_signature = function[signature_property].find { |sig| sig['kindString'] == kind_string }
      function_object = {
        'name' => function['name'],
        'text' => get_desc(function_signature['comment']),
        'parameters' => [],
        'returns' => get_type(function_signature['type'], 'function'),
        'returns text' => get_returns(function_signature['comment'])
      }
      function_object['type_parameters'] = expand_type_parameters(function_signature['typeParameter']) if function_signature.key?('typeParameter')
      unless function_signature['parameters'].nil?
        function_signature['parameters'].each do |parameter|
          parameter_object = {
            'name' => parameter['name'],
            'type' => get_type(parameter['type'], 'parameter'),
            'text' => get_desc(parameter['comment'])
          }
          function_object['parameters'].push(parameter_object)
        end
      end
      alphabetize(function_object['parameters'])
      functions_expanded.push(function_object)
    end
    alphabetize(functions_expanded)
    functions_expanded
  end

  def expand_initializer(function)
    function_signature = function['signatures'].find { |sig| sig['kindString'] == 'Call signature' }
    function_object = {
      'name' => function['name'],
      'text' => get_desc(function_signature['comment']),
      'parameters' => [],
      'returns' => get_controller_methods(function_signature['type']),
      'returns text' => get_returns(function_signature['comment'])
    }
    function_object['type_parameters'] = expand_type_parameters(function_signature['typeParameter']) if function_signature.key?('typeParameter')
    unless function_signature['parameters'].nil?
      function_signature['parameters'].each do |parameter|
        parameter_object = {
          'name' => parameter['name'],
          'type' => get_type(parameter['type'], 'parameter'),
          'text' => get_desc(parameter['comment'])
        }
        function_object['parameters'].push(parameter_object)
      end
    end
    alphabetize(function_object['parameters'])
    function_object
  end

  def get_controller_methods(type_object)
    functions = []
    accessors = []
    type_object['declaration']['children'].each do |method|
      functions.push(method) if method['kindString'] == 'Function'
      accessors.push(method) if method['kindString'] == 'Accessor'
    end
    {
      'methods' => expand_functions(functions),
      'accessors' => expand_accessors(accessors)
    }
  end

  def expand_actions(actions)
    actions.map do |action|
      {
        'name' => action['name'],
        'text' => get_desc(action['comment']),
        'parameters' => get_action_parameters(action)
      }
    end
  end

  def get_action_parameters(action)
    return [] if action['comment'].nil? || action['comment']['tags'].nil?

    action['comment']['tags'].map do |param|
      {
        'name' => param['param'],
        'text' => get_desc(param)
      }
    end
  end

  def expand_type_parameters(type_parameters)
    type_parameters.map do |param|
      obj = {
        'name' => param['name']
      }
      obj['extends'] = get_type(param['type'], 'generic')
      obj
    end
  end

  # Engine

  def parse_engine
    engine_modules = []
    @config['engine']['sources'].each { |source| engine_modules.concat(get_modules(source)) }
    {
      'headless_engine' => get_headless_engine(engine_modules),
      'types' => expand_types(get_entities_from_modules_by_kind_string(engine_modules, 'Type alias')),
      'interfaces' => expand_interfaces(get_entities_from_modules_by_kind_string(engine_modules, 'Interface')),
      'functions' => expand_functions(get_entities_from_modules_by_kind_string(engine_modules, 'Function'))
    }
  end

  def get_headless_engine(engine_modules)
    headless_engine = get_entity_from_modules_by_name(engine_modules, 'HeadlessEngine')
    {
      'constructor' => expand_constructor(get_entity_from_module_by_name(headless_engine, 'constructor')),
      'properties' => expand_types(get_entities_from_module_by_kind_string(headless_engine, 'Property')),
      'methods' => expand_functions(get_entities_from_module_by_kind_string(headless_engine, 'Method')),
      'accessors' => expand_accessors(get_entities_from_module_by_kind_string(headless_engine, 'Accessor'))
    }
  end

  # Controllers

  def parse_controllers
    @config['controllers'].map do |controller_config|
      controller_modules = []
      controller_config['sources'].each { |source| controller_modules.concat(get_modules(source)) }
      {
        'name' => controller_config['name'],
        'initializer' => expand_initializer(get_entity_from_modules_by_name(controller_modules, "build#{controller_config['name']}")),
        'types' => expand_types(get_entities_from_modules_by_kind_string(controller_modules, 'Type alias')),
        'interfaces' => expand_interfaces(get_entities_from_modules_by_kind_string(controller_modules, 'Interface')),
        'functions' => expand_functions(get_entities_from_modules_by_kind_string(controller_modules, 'Function').reject { |function| function['name'] == "build#{controller_config['name']}"}),
        'enums' => expand_enums(get_entities_from_modules_by_kind_string(controller_modules, 'Enumeration'))
      }
    end
  end

  # Actions

  def parse_actions
    @config['actions']['sections'].map do |section|
      section_modules = []
      section['sources'].each { |source| section_modules.concat(get_modules(source)) }
      actions = get_entities_from_modules_by_kind_string(section_modules, 'Variable')
      actions.concat(get_entities_from_modules_by_kind_string(section_modules, 'Function'))
      actions.select! do |action|
        action['flags'].key?('isExported') && action['flags']['isExported']
      end
      {
        'section' => section['name'],
        'actions' => expand_actions(actions)
      }
    end
  end
end

def main
  File.write(OUTPUT_PATH, JSON.pretty_generate(DocJsonParser.new(DOC_GEN_PATH, CONFIG_PATH).parse))
end

main

File: packages/headless/DocJsonParser.ts
import {readFileSync, writeFileSync} from 'fs';

const DOC_GEN_PATH = './packages/headless/dist/doc.json';
const CONFIG_PATH = './packages/headless/docs.config.json';
const OUTPUT_PATH = './packages/headless/dist/parsed_doc.json';

type Config = {
  engine: {
    sources: string[];
  };
  actions: {
    sections: {
      name: string;
      sources: string[];
    }[];
  };
  controllers: {
    name: string;
    sources: string[];
  }[];
};

type EntityType = {
  type: string;
  name?: string;
  value?: string;
  elementType?: EntityType;
  types?: EntityType[];
  typeArguments?: EntityType[];
  constraint?: EntityType;
  default?: EntityType;
  declaration?: Entity;
};

type DocComment = {
  shortText?: string;
  text?: string;
  returns?: string;
  tags?: {
    tag: string;
    text?: string;
    param?: string;
  }[];
};

type Entity = {
  name: string;
  kindString: string;
  flags?: {
    isExported?: boolean;
    isOptional?: boolean;
  };
  comment?: DocComment;
  type?: EntityType;
  parameters?: Entity[];
  typeParameter?: Entity[];
  children?: Entity[];
  signatures?: Entity[];
  getSignature?: Entity[];
};

type Module = {
  originalName: string;
  children: Entity[];
};

type DocGen = {
  children: Module[];
};

class DocJsonParser {
  private docgen: DocGen;
  private config: Config;

  constructor(docgenPath: string, configPath: string) {
    this.docgen = JSON.parse(readFileSync(docgenPath, 'utf8'));
    this.config = JSON.parse(readFileSync(configPath, 'utf8'));
  }

  parse() {
    return {
      engine: this.parseEngine(),
      controllers: this.parseControllers(),
      actions: this.parseActions(),
    };
  }

  private parseEngine() {
    let engineModules: Module[] = [];
    for (const source of this.config.engine.sources) {
      engineModules = engineModules.concat(this.getModules(source));
    }
    return {
      headless_engine: this.getHeadlessEngine(engineModules),
      types: this.expandTypes(
        this.getFromModulesByKindString(engineModules, 'Type alias')
      ),
      interfaces: this.expandInterfaces(
        this.getFromModulesByKindString(engineModules, 'Interface')
      ),
      functions: this.expandFunctions(
        this.getFromModulesByKindString(engineModules, 'Function')
      ),
    };
  }

  private parseControllers() {
    return this.config.controllers.map((controllerConfig) => {
      let controllerModules: Module[] = [];
      for (const source of controllerConfig.sources) {
        controllerModules = controllerModules.concat(this.getModules(source));
      }

      return {
        name: controllerConfig.name,
        initializer: this.expandInitializer(
          this.getFromModulesByName(
            controllerModules,
            `build${controllerConfig.name}`
          )
        ),
        types: this.expandTypes(
          this.getFromModulesByKindString(controllerModules, 'Type alias')
        ),
        interfaces: this.expandInterfaces(
          this.getFromModulesByKindString(controllerModules, 'Interface')
        ),
        functions: this.expandFunctions(
          this.getFromModulesByKindString(controllerModules, 'Function').filter(
            (value) => {
              return value.name !== `build${controllerConfig.name}`;
            }
          )
        ),
        enums: this.expandEnums(
          this.getFromModulesByKindString(controllerModules, 'Enumeration')
        ),
      };
    });
  }

  private parseActions() {
    return this.config.actions.sections.map((section) => {
      let sectionModules: Module[] = [];
      for (const source of section.sources) {
        sectionModules = sectionModules.concat(this.getModules(source));
      }
      let actions: Entity[] = this.getFromModulesByKindString(
        sectionModules,
        'Variable'
      );
      actions = actions.concat(
        this.getFromModulesByKindString(sectionModules, 'Function')
      );
      actions = actions.filter((action) => {
        return (
          typeof action.flags.isExported !== undefined &&
          action.flags.isExported === true
        );
      });
      return {
        section: section.name,
        actions: this.expandActions(actions),
      };
    });
  }

  private getHeadlessEngine(engineModules: Module[]) {
    const headlessEngine = this.getFromModulesByName(
      engineModules,
      'HeadlessEngine'
    );
    return {
      text: this.getDesc(headlessEngine.comment),
      constructor: this.expandConstructor(
        this.getFromModuleByName(headlessEngine, 'constructor')
      ),
      properties: this.expandTypes(
        this.getFromModuleByKindString(headlessEngine, 'Property')
      ),
      methods: this.expandFunctions(
        this.getFromModuleByKindString(headlessEngine, 'Method')
      ),
      accessors: this.expandAccessors(
        this.getFromModuleByKindString(headlessEngine, 'Accessor')
      ),
    };
  }

  // ----------------Retrieving stuff from the docgen------------------------

  private getModules(modulePath: string) {
    return this.docgen.children.filter((child) => {
      return child.originalName.includes(modulePath);
    });
  }

  private getFromModulesByName(modules: Module[] | Entity[], name: string) {
    for (const mod of modules) {
      const entity = this.getFromModuleByName(mod, name);
      if (entity) return entity;
    }
    return null;
  }

  private getFromModuleByName(mod: Module | Entity, name: string) {
    return mod.children.find((entity) => {
      return entity.name === name;
    });
  }

  private getFromModulesByKindString(
    modules: Module[] | Entity[],
    kindString: string
  ) {
    const entities: Entity[] = [];
    for (const mod of modules) {
      entities.concat(this.getFromModuleByKindString(mod, kindString));
    }
    return entities;
  }

  private getFromModuleByKindString(mod: Module | Entity, kindString: string) {
    return mod.children.filter((entity) => {
      return entity.kindString === kindString;
    });
  }

  private getDesc(comment: DocComment) {
    if (!comment) return '';
    if (!comment.shortText && !comment.text) return '';

    if (comment.shortText && comment.text)
      return `${comment.shortText}\n\n${comment.text}`;

    return comment.shortText ? comment.shortText : comment.text;
  }

  private getReturns(comment: DocComment) {
    if (!comment) return '';
    return comment.returns ? comment.returns : '';
  }

  //-------------------------Expanding the various entities---------------------

  private alphabetize(objectArray: {name: string}[]) {
    return objectArray.sort((a, b) => {
      return a.name.localeCompare(b.name);
    });
  }

  private expandFunctions(functions: Entity[]) {
    return this.expandAnyFunctions(functions, 'Call signature', 'signatures');
  }

  private expandAccessors(functions: Entity[]) {
    return this.expandAnyFunctions(functions, 'Get signature', 'getSignature');
  }

  private expandConstructor(func: Entity) {
    return this.expandAnyFunctions(
      [func],
      'Constructor signature',
      'signatures'
    )[0];
  }

  private expandTypes(types: Entity[]) {
    const typesExpanded = types.map((type) => {
      const typeObject = {
        name: type.name,
        type: this.getType(type.type, 'type'),
        text: this.getDesc(type.comment),
        type_parameters: [],
      };
      if (type.typeParameter)
        typeObject.type_parameters = this.expandTypeParameters(
          type.typeParameter
        );
      return typeObject;
    });
    return this.alphabetize(typesExpanded);
  }

  private expandInterfaces(interfaces: Entity[]) {
    const interfacesExpanded = interfaces.map((iface) => {
      const interfaceObject = {
        name: iface.name,
        text: this.getDesc(iface.comment),
        properties: iface.children.map((property) => {
          const propertyObject = {
            name: property.name,
            type: this.getType(property.type, 'interface'),
            text: this.getDesc(property.comment),
            optional: false,
          };
          if (property.flags.isOptional) propertyObject.optional = true;
          return propertyObject;
        }),
        type_parameters: [],
      };
      if (iface.typeParameter)
        interfaceObject.type_parameters = this.expandTypeParameters(
          iface.typeParameter
        );
      return interfaceObject;
    });
    return this.alphabetize(interfacesExpanded);
  }

  private expandEnums(enums: Entity[]) {
    const enumsExpanded = enums.map((enumeration) => {
      return {
        name: enumeration.name,
        text: this.getDesc(enumeration.comment),
        members: this.alphabetize(
          enumeration.children.map((member) => {
            return {
              name: member.name,
              text: this.getDesc(member.comment),
            };
          })
        ),
      };
    });
    return this.alphabetize(enumsExpanded);
  }

  private expandAnyFunctions(
    functions: Entity[],
    kindString: string,
    signatureProperty: string
  ) {
    const functionsExpanded = functions.map((func) => {
      const functionSignature = func[signatureProperty].find((sig) => {
        return sig.kindString === kindString;
      });
      return {
        name: func.name,
        text: this.getDesc(functionSignature.comment),
        parameters: functionSignature.parameters
          ? functionSignature.parameters.map((param) => {
              return {
                name: param.name,
                type: this.getType(param.type, 'parameter'),
                text: this.getDesc(param.comment),
              };
            })
          : [],
        returns: this.getType(functionSignature.type, 'function'),
        returns_text: this.getReturns(functionSignature.comment),
        type_parameters: functionSignature.typeParameter
          ? this.expandTypeParameters(functionSignature.typeParameter)
          : [],
      };
    });
    return this.alphabetize(functionsExpanded);
  }

  private expandInitializer(func: Entity) {
    const functionSignature = func.signatures.find((sig) => {
      return sig.kindString === 'Call signature';
    });
    return {
      name: func.name,
      text: this.getDesc(functionSignature.comment),
      parameters: functionSignature.parameters.map((param) => {
        return {
          name: param.name,
          type: this.getType(param.type, 'parameter'),
          text: this.getDesc(param.comment),
        };
      }),
      returns: this.getControllerMethods(functionSignature.type),
      returns_text: this.getReturns(functionSignature.comment),
      type_parameters: functionSignature.typeParameter
        ? this.expandTypeParameters(functionSignature.typeParameter)
        : [],
    };
  }

  private getControllerMethods(typeObject: EntityType) {
    const functions: Entity[] = [];
    const accessors: Entity[] = [];
    for (const method of typeObject.declaration.children) {
      if (method.kindString === 'Function') functions.push(method);
      if (method.kindString === 'Accessor') accessors.push(method);
    }
    return {
      methods: this.expandFunctions(functions),
      accessors: this.expandAccessors(accessors),
    };
  }

  private expandActions(actions: Entity[]) {
    return actions.map((action) => {
      return {
        name: action.name,
        text: this.getDesc(action.comment),
        parameters: this.getActionParameters(action),
      };
    });
  }

  private getActionParameters(action: Entity) {
    if (!action.comment || !action.comment.tags) return [];

    return action.comment.tags.map(function (param) {
      return {
        name: param.param,
        text: this.getDesc(param),
      };
    });
  }

  private expandTypeParameters(typeParameters: Entity[]) {
    return typeParameters.map((param) => {
      return {
        name: param.name,
        extends: this.getType(param.type, 'generic'),
      };
    });
  }

  // -----------------Type Resolver-------------------------------------

  private getType(typeObject: EntityType, section: string): object {
    if (!typeObject) return {};
    if (this.isType(typeObject, 'stringLiteral'))
      return this.processStringLiteral(typeObject);
    if (this.isType(typeObject, 'intrinsic'))
      return this.processIntrinsicType(typeObject);
    if (this.isType(typeObject, 'array'))
      return this.processArrayType(typeObject);
    if (this.isType(typeObject, 'union'))
      return this.processUnionType(typeObject, section);
    if (this.isType(typeObject, 'intersection'))
      return this.processIntersectionType(typeObject);
    if (this.isType(typeObject, 'reference'))
      return this.processReferenceType(typeObject, section);
    if (this.isType(typeObject, 'typeParameter'))
      return this.processTypeParameterType(typeObject);
    if (this.isType(typeObject, 'reflection'))
      return this.processReflectionType(typeObject);

    return {};
  }

  private isType(typeObject: EntityType, targetType: string) {
    return typeObject.type === targetType;
  }

  private processStringLiteral(typeObject: EntityType) {
    return {
      type: typeObject.type,
      value: typeObject.value,
    };
  }

  private processIntrinsicType(typeObject: EntityType) {
    return {
      type: typeObject.type,
      name: typeObject.name,
    };
  }

  private processArrayType(typeObject: EntityType) {
    return {
      type: 'array',
      arrayOf: this.getType(typeObject.elementType, 'array'),
    };
  }

  private processUnionType(typeObject: EntityType, section: string) {
    const processedType = {
      type: 'union',
      unionOf: typeObject.types.map((type) => {
        return this.getType(type, 'union');
      }),
      optional: false,
    };
    if (
      section === 'interface' &&
      processedType.unionOf.includes({type: 'intrinsic', name: 'undefined'})
    ) {
      processedType.optional = true;
      processedType.unionOf = processedType.unionOf.filter((type) => {
        return type !== {type: 'intrinsic', name: 'undefined'};
      });
    }
    return processedType;
  }

  private processIntersectionType(typeObject: EntityType) {
    return {
      type: 'intersection',
      intersectionOf: typeObject.types.map((type) => {
        return this.getType(type, 'intersection');
      }),
    };
  }

  private processReferenceType(typeObject: EntityType, section: string) {
    const processedType = {
      type: 'reference',
      name: typeObject.name,
      type_parameters: [],
    };
    if (typeObject.typeArguments) {
      processedType.type_parameters = typeObject.typeArguments.map(
        (typeArg) => {
          return this.getType(typeArg, section);
        }
      );
    }
    return processedType;
  }

  private processTypeParameterType(typeObject: EntityType) {
    const processedType = {
      type: 'type_parameter',
      name: typeObject.name,
      extends: {},
      default: {},
    };
    if (typeObject.constraint)
      processedType.extends = this.getType(typeObject.constraint, 'generic');
    if (typeObject.default)
      processedType.default = this.getType(typeObject.default, 'generic');
    return processedType;
  }

  private processReflectionType(typeObject: EntityType) {
    if (typeObject.declaration.children) {
      return {
        type: 'object',
        properties: typeObject.declaration.children.map((property) => {
          return {
            name: property.name,
            type: this.getType(property.type, 'object'),
          };
        }),
      };
    } else if (typeObject.declaration.signatures) {
      const processedType = {
        type: 'function',
        parameters: [],
        returns: this.getType(
          typeObject.declaration.signatures[0].type,
          'function'
        ),
      };
      if (typeObject.declaration.signatures[0].parameters) {
        for (const param of typeObject.declaration.signatures[0].parameters) {
          processedType.parameters.push({
            name: param.name,
            type: this.getType(param.type, 'parameter'),
          });
        }
      }
      return processedType;
    } else {
      return {};
    }
  }
}

const parser = new DocJsonParser(DOC_GEN_PATH, CONFIG_PATH);
writeFileSync(OUTPUT_PATH, JSON.stringify(parser.parse(), null, 2));
